zkvm/src/signature/signer.rs

```diff
@@ -165,28 +155,15 @@ impl PartyAwaitingShares {
         Signature { s, R: self.R }
     }
 
-    pub fn receive_shares(
-        self,
-        shares: Vec<Scalar>,
-        pubkeys: Vec<VerificationKey>,
-    ) -> Result<Signature, VMError> {
+    pub fn receive_shares(self, shares: Vec<Scalar>) -> Result<Signature, VMError> {
         // Check that all shares are valid
-        for (i, s_i) in shares.iter().enumerate() {
-            let S_i = s_i * RISTRETTO_BASEPOINT_POINT;
-            let X_i = pubkeys[i];
-            let R_i = self.nonce_commitments[i].0;
-
-            // Make a_i = H(L, X_i)
-            let a_i = self.multikey.factor_for_key(&VerificationKey(X_i.0));
-
-            let X_i = X_i.0.decompress().ok_or(VMError::InvalidPoint)?;
-
-            // Check that S_i = R_i + c * a_i * X_i
-            if S_i != R_i + self.c * a_i * X_i {
-                return Err(VMError::ShareError { index: i });
-            }
-        }
+        let validated_shares = self
+            .counterparties
+            .iter()
```

Comment:
Help! When I try to take ownership of `self.counterparties` by using `into_iter()` instead of `iter`, I then no longer have ownership over `self` in future steps. This prevents me from using `self.c`, `self.multikey` inside the map, as well as prevents me from calling `self.receive_trusted_shares(...)`. 

So, for now I am changing `CounterpartyCommitted::sign(...)` to take `&self` instead of `self`. This is not great (since then you can call this function on the struct multiple times, which is not *dangerous* but is just incorrect behavior). Is there a way to take ownership of `self.counterparties` without taking ownership of all of `self`, and therefore being able to take ownership of other fields of `self`?

(To get the errors I'm talking about, remove the `&` from `&self` in the inputs of `CounterpartyCommitted::sign(...)`.)

Reply:
Simple solution: std::mem::swap with an empty vector

Reply:
The issue here is not that you need `self.c` and `self.multikey` later - you can move out fields in arbitrary order. The issue is that the `map`'s closure captures `self`, not its individual fields like it seems at first.

So to avoid this problem, you'd need to first move out `self.c` and `self.multikey` in local variables outside the closure, and then capture them in the closure, so `self` is not referenced by anyone else. This should allow you to do `self.counterparties.into_iter()` and have clean move semantics throughout.

Reply:
Of course, there's an RFC for that already: https://github.com/rust-lang/rfcs/pull/2229 + https://github.com/rust-lang/rust/issues/53488 

Meanwhile, this is what i found:
* http://smallcultfollowing.com/babysteps/blog/2018/04/24/rust-pattern-precise-closure-capture-clauses/
* https://github.com/rust-lang/rust/issues/19004
* https://stevedonovan.github.io/rustifications/2018/08/18/rust-closures-are-hard.html

Especially a neat trick in the first article:

<img width="903" alt="image" src="https://user-images.githubusercontent.com/698/54648496-b62b0080-4a63-11e9-8597-7880ae68d39a.png">

TL;DR: you move out those fields into local variables inside the block (not closure) located where the closure needs to be, which returns a closure that uses those local variables. So it feels like it the closure with "explicit capture annotations" using standard syntax features.

Reply:
I tried what you suggested (moving out the fields into local variables), but the problem is that after the map, I still want to be able to call `self.receive_trusted_shares(...)`. And I can't do that because the map's closure captures `self`, so I no longer have ownership of `self` to call `self.receive_trusted_shares(...)` on.

I will read into the articles you linked, maybe those will have something! Otherwise it might just make sense to copy the logic in `receive_trusted_shares()` and put it in `receive_shares()` to avoid the problem of calling `self.receive_trusted_shares` without ownership of `self`.

Reply:
@cathieyun don't call `receive_trusted_shares` :-) it's a one-liner that sums up all shares, so you can just do that directly in the `receive_shares`.

Reply:
moreover, it's more efficient that collecting into Result<Vec<_>>, if you simply end the giant iterator with .sum(). And the sum over iteration of Results is defined in stdlib to return Result of the sum!

<img width="797" alt="image" src="https://user-images.githubusercontent.com/698/54700822-f71d2680-4af0-11e9-9e08-9965fc9c5b6e.png">

Reply:
That's awesome, I had no idea you could do a sum over iter of Results! Nice :)